README.md
## Circuit Breaker
[![Gem Version](https://img.shields.io/gem/v/circuit_breaker-ruby.svg?style=flat)](http://rubygems.org/gems/circuit_breaker-ruby)
[![Gem Downloads](https://img.shields.io/gem/dt/circuit_breaker-ruby.svg?style=flat)](http://rubygems.org/gems/circuit_breaker-ruby)
[![Build Status](https://travis-ci.org/scripbox/circuit_breaker-ruby.svg?style=flat&branch=master)](https://travis-ci.org/scripbox/circuit_breaker-ruby)

  A circuit breaker which terminates the connection or block of code from executing when it reaches the timeout. It helps in preventing blocking requests from slowing down your server.

## How it works

Refer the following diagram:

![CircuitBreaker](https://raw.githubusercontent.com/scripbox/circuit_breaker-ruby/master/circuit_breaker.png)

## Getting started

  Add following line to your Gemfile:
  ```ruby
  gem 'circuit_breaker-ruby', '~> 0.1.3'
  ```

## Usage

  ```ruby
  circuit_breaker = CircuitBreaker::Shield.new(
    invocation_timeout: 1,
    failure_threshold: 2,
    failure_threshold_percentage: 0.2,
    retry_timeout: 10
  )
  circuit_breaker.protect { sleep(10) }
  ```

## Running tests

  ```ruby
  bundle exec rake
  ```

## Configuration

Add the following configuration to `config/initializers/circuit_breaker.rb`. These are the default values.

  ```ruby
  CircuitBreaker.configure do |cb|
    cb.failure_threshold = 10
    cb.failure_threshold_percentage = 0.5
    cb.invocation_timeout = 10
    cb.retry_timeout = 60
  end
  ```

## Contributing

If you have any issues with circuit_breaker-ruby,
or feature requests,
please [add an issue](https://github.com/scripbox/circuit_breaker-ruby/issues) on GitHub
or fork the project and send a pull request.
Please include passing specs with all pull requests.

## Copyright and License

Copyright (c) 2019, Scripbox.

circuit_breaker-ruby source code is licensed under the [MIT License](MIT-LICENSE).

## References

[CircuitBreaker](https://martinfowler.com/bliki/CircuitBreaker.html) - [Martin Fowler](https://en.wikipedia.org/wiki/Martin_Fowler_(software_engineer))


CHANGELOG.md
# Changelog for v0.1.x

## v0.1.3 (2019-03-12)

### Bug fixes

  * Total count is not getting incremented incase of exception is raised
  * total_count reset to 0 when state is in half_open
  * reached_failure_threshold? returns only true once it is set to true


spec/spec_helper.rb
require 'rubygems'
require 'rspec'

RSpec.configure do |config|
  config.tty = true
  config.color = true
end


spec/circuit_breaker_spec.rb
require 'spec_helper'
require 'circuit_breaker-ruby'

describe CircuitBreaker do
  let(:circuit_breaker_instance) do
    CircuitBreaker.new(invocation_timeout: 1, retry_timeout: 1, failure_threshold: 1)
  end

  it 'goes to closed state' do
    circuit_breaker_instance.call { sleep(0.1) }

    expect(circuit_breaker_instance.send :state).to be(CircuitBreaker::States::CLOSED)
  end

  it 'goes to open state' do
    no_of_tries = circuit_breaker_instance.failure_threshold * 2
    no_of_failures = no_of_tries * CircuitBreaker::FAILURE_THRESHOLD_PERCENTAGE
    no_of_success = no_of_tries - no_of_failures
    no_of_success.to_i.times { circuit_breaker_instance.call { sleep(0.1) } }
    no_of_failures.to_i.times { circuit_breaker_instance.call { sleep(2) } }

    expect(circuit_breaker_instance.send :state).to be(CircuitBreaker::States::OPEN)
  end

  it 'goes to half open state' do
    no_of_tries = circuit_breaker_instance.failure_threshold * 2
    no_of_failures = no_of_tries * CircuitBreaker::FAILURE_THRESHOLD_PERCENTAGE
    no_of_success = no_of_tries - no_of_failures
    no_of_success.to_i.times { circuit_breaker_instance.call { sleep(0.1) } }
    no_of_failures.to_i.times { circuit_breaker_instance.call { sleep(2) } }
    sleep(1)

    expect(circuit_breaker_instance.send :state).to be(CircuitBreaker::States::HALF_OPEN)
  end

  it 'raises CircuitBreaker::Open' do
     no_of_tries = circuit_breaker_instance.failure_threshold * 2
    no_of_failures = no_of_tries * CircuitBreaker::FAILURE_THRESHOLD_PERCENTAGE
    no_of_success = no_of_tries - no_of_failures
    no_of_success.to_i.times { circuit_breaker_instance.call { sleep(0.1) } }

    expect { (no_of_failures.to_i + 1).times { circuit_breaker_instance.call { sleep(2) } } }.to(
      raise_error(CircuitBreaker::Open)
    )
  end
end


spec/circuit_breaker/shield_spec.rb
require 'spec_helper'
require 'circuit_breaker-ruby'
require 'timecop'

describe CircuitBreaker::Shield do
  context 'with no failures' do
    it 'goes to closed state' do
      circuit_breaker_shield = CircuitBreaker::Shield.new(
        invocation_timeout: 1,
        retry_timeout: 1,
        failure_threshold: 1
      )

      circuit_breaker_shield.protect { sleep(0.1) } # succeed once
      expect(circuit_breaker_shield.total_count).to eql(1)
      expect(circuit_breaker_shield.failure_count).to eql(0)
      expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::CLOSED)
    end
  end

  context 'with failures less than threshold' do
    it 'goes to closed state' do
      retry_timeout = 1

      circuit_breaker_shield = CircuitBreaker::Shield.new(
        invocation_timeout: 1,
        retry_timeout: retry_timeout,
        failure_threshold: 2
      )

      circuit_breaker_shield.protect { sleep(0.1) } # succeed once
      expect(circuit_breaker_shield.total_count).to eql(1)
      expect(circuit_breaker_shield.failure_count).to eql(0)

      expect {circuit_breaker_shield.protect { sleep(1.1) }}.to raise_error(CircuitBreaker::TimeoutError) # fail once
      expect(circuit_breaker_shield.total_count).to eql(2)
      expect(circuit_breaker_shield.failure_count).to eql(1)
      expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::CLOSED)
    end
  end

  context 'with failures more than threshold' do
    context 'within retry_timeout' do
      it 'goes to open state' do
        circuit_breaker_shield = CircuitBreaker::Shield.new(
          invocation_timeout: 1,
          retry_timeout: 1,
          failure_threshold: 1
        )

        circuit_breaker_shield.protect { sleep(0.1) } # succeed once
        expect(circuit_breaker_shield.total_count).to eql(1)
        expect(circuit_breaker_shield.failure_count).to eql(0)

        expect { circuit_breaker_shield.protect { sleep(1.1) } }.to raise_error(CircuitBreaker::TimeoutError) # fail once
        expect(circuit_breaker_shield.total_count).to eql(2)
        expect(circuit_breaker_shield.failure_count).to eql(1)

        expect { circuit_breaker_shield.protect { sleep(1.1) } }.to raise_error(CircuitBreaker::Open) # fail twice
        expect(circuit_breaker_shield.total_count).to eql(2)
        expect(circuit_breaker_shield.failure_count).to eql(1)

        expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::OPEN)
      end
    end

    context 'after retry_timeout' do
      it 'goes to half open state' do
        retry_timeout = 1

        circuit_breaker_shield = CircuitBreaker::Shield.new(
          invocation_timeout: 1,
          retry_timeout: retry_timeout,
          failure_threshold: 1
        )

        circuit_breaker_shield.protect { sleep(0.1) } # succeed once
        expect(circuit_breaker_shield.total_count).to eql(1)
        expect(circuit_breaker_shield.failure_count).to eql(0)

        expect {circuit_breaker_shield.protect { sleep(1.1) }}.to raise_error(CircuitBreaker::TimeoutError) # fail once
        expect(circuit_breaker_shield.total_count).to eql(2)
        expect(circuit_breaker_shield.failure_count).to eql(1)


        Timecop.freeze(Time.now + retry_timeout) do
          expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::HALF_OPEN)
        end
      end
    end

    context 'when circuit is in half-open state' do
      it 'goes to open state' do
        retry_timeout = 1

        circuit_breaker_shield = CircuitBreaker::Shield.new(
          invocation_timeout: 1,
          retry_timeout: retry_timeout,
          failure_threshold: 1
        )

        circuit_breaker_shield.protect { sleep(0.1) } # succeed once
        expect(circuit_breaker_shield.total_count).to eql(1)
        expect(circuit_breaker_shield.failure_count).to eql(0)

        expect {circuit_breaker_shield.protect { sleep(1.1) }}.to raise_error(CircuitBreaker::TimeoutError) # fail once
        expect(circuit_breaker_shield.total_count).to eql(2)
        expect(circuit_breaker_shield.failure_count).to eql(1)

        Timecop.freeze(Time.now + retry_timeout) do
          expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::HALF_OPEN)

          expect {circuit_breaker_shield.protect { sleep(1.1) }}.to raise_error(CircuitBreaker::TimeoutError)
          expect(circuit_breaker_shield.total_count).to eql(3)
          expect(circuit_breaker_shield.failure_count).to eql(2)

          expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::OPEN)
        end
      end
    end

    context 'when circuit is in open state and goes to closed state' do
      it 'remains in closed state until it reaches failure_threshold' do
        retry_timeout = 3

        circuit_breaker_shield = CircuitBreaker::Shield.new(
          invocation_timeout: 1,
          retry_timeout: retry_timeout,
          failure_threshold: 2
        )

        circuit_breaker_shield.protect { sleep(0.1) } # succeed once
        expect(circuit_breaker_shield.total_count).to eql(1)
        expect(circuit_breaker_shield.failure_count).to eql(0)

        expect {circuit_breaker_shield.protect { sleep(1.1) }}.to raise_error(CircuitBreaker::TimeoutError) # fail once
        expect(circuit_breaker_shield.total_count).to eql(2)
        expect(circuit_breaker_shield.failure_count).to eql(1)
        expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::CLOSED)

        expect {circuit_breaker_shield.protect { sleep(1.1) }}.to raise_error(CircuitBreaker::TimeoutError) # fail twice
        expect(circuit_breaker_shield.total_count).to eql(3)
        expect(circuit_breaker_shield.failure_count).to eql(2)
        expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::OPEN)

        Timecop.freeze(Time.now + retry_timeout) do
          circuit_breaker_shield.protect { sleep(0.1) } # succeed once
          expect(circuit_breaker_shield.total_count).to eql(1)
          expect(circuit_breaker_shield.failure_count).to eql(0)
          expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::CLOSED)

          expect {circuit_breaker_shield.protect { sleep(1.1) }}.to raise_error(CircuitBreaker::TimeoutError) # fail once
          expect(circuit_breaker_shield.total_count).to eql(2)
          expect(circuit_breaker_shield.failure_count).to eql(1)
          expect(circuit_breaker_shield.send(:state)).to be(CircuitBreaker::Shield::States::CLOSED)
        end
      end
    end
  end

  context '#protect' do
    it 'update invocation_timeout in config' do
      circuit_breaker_shield = CircuitBreaker::Shield.new(
        invocation_timeout: 1,
        retry_timeout: 1,
        failure_threshold: 1
      )

      circuit_breaker_shield.protect(invocation_timeout: 20) {}

      expect(circuit_breaker_shield.invocation_timeout).to be(20)
    end

    it 'invokes callback on success' do
      callback = proc { 'Success' }
      circuit_breaker_shield = CircuitBreaker::Shield.new(
        invocation_timeout: 1,
        retry_timeout: 1,
        failure_threshold: 1,
        callback: callback
      )

      expect(callback).to receive(:call).and_return('Success')

      circuit_breaker_shield.protect { { response: 'Dummy response' } }
    end

    context 'with timeout error' do
      it 're-raises timeout error' do
        circuit_breaker_shield = CircuitBreaker::Shield.new(
          invocation_timeout: 1,
          retry_timeout: 1,
          failure_threshold: 1
        )

        invocation_block = proc{ raise Timeout::Error }

        expect {circuit_breaker_shield.protect(&invocation_block)}.to raise_error(CircuitBreaker::TimeoutError)
      end
    end
  end
end


lib/circuit_breaker-ruby/config.rb
module CircuitBreaker
  class Config
    FAILURE_THRESHOLD = 10
    FAILURE_THRESHOLD_PERCENTAGE = 0.5
    INVOCATION_TIMEOUT = 10
    RETRY_TIMEOUT = 60

    UPDATABLE = [
      :invocation_timeout,
      :failure_threshold,
      :failure_threshold_percentage,
      :retry_timeout
    ]

    attr_accessor :invocation_timeout, :failure_threshold, :failure_threshold_percentage, :retry_timeout

    def initialize
      self.failure_threshold = FAILURE_THRESHOLD
      self.failure_threshold_percentage = FAILURE_THRESHOLD_PERCENTAGE
      self.invocation_timeout = INVOCATION_TIMEOUT
      self.retry_timeout = RETRY_TIMEOUT
    end

    def self.update(klass, options)
      (UPDATABLE & options.keys).each do |variable|
        klass.instance_variable_set("@#{variable}", options[variable])
      end
    end
  end
end


lib/circuit_breaker-ruby/shield.rb
module CircuitBreaker
  class Shield
    module States
      OPEN = :open
      CLOSED = :closed
      HALF_OPEN = :half_open
    end

    attr_reader :invocation_timeout,
                :retry_timeout,
                :failure_threshold,
                :failure_threshold_percentage,
                :total_count,
                :failure_count

    def initialize(**options)
      @failure_count = 0
      @total_count = 0
      @failure_threshold = options.fetch(:failure_threshold, config.failure_threshold)
      @failure_threshold_percentage = options.fetch(:failure_threshold_percentage, config.failure_threshold_percentage)
      @invocation_timeout = options.fetch(:invocation_timeout, config.invocation_timeout)
      @retry_timeout = options.fetch(:retry_timeout, config.retry_timeout)
      @callback = options[:callback]
    end

    def config
      CircuitBreaker.config
    end

    def update_config!(options)
      CircuitBreaker::Config.update(self, options)
    end

    def protect(options = {}, &block)
      update_config!(options)

      case prev_state = state
      when States::CLOSED, States::HALF_OPEN
        connect(&block)
      when States::OPEN
        raise CircuitBreaker::Open
      end
    end

    private

    def state
      if reached_failure_threshold? && reached_failure_threshold_percentage? && reached_retry_timeout?
        States::HALF_OPEN
      elsif reached_failure_threshold? && reached_failure_threshold_percentage?
        States::OPEN
      else
        States::CLOSED
      end
    end

    def reached_failure_threshold?
      (failure_count >= failure_threshold)
    end

    def reached_failure_threshold_percentage?
      (total_count.nonzero? && (failure_count.to_f / total_count.to_f) >= failure_threshold_percentage)
    end

    def reached_retry_timeout?
      (Time.now - @last_failure_time) > @retry_timeout
    end

    def reset
      @failure_count = 0
      @total_count = 0
      @state = States::CLOSED
    end

    def connect(&block)
      begin
        result = nil
        ::Timeout::timeout(invocation_timeout) do
          start_time = Time.now
          result = block.call
          duration = Time.now - start_time
          invoke_callback(result, duration: duration)
          reset
        end
      rescue ::Timeout::Error => e
        record_failure
        invoke_callback
        raise CircuitBreaker::TimeoutError
      ensure
        increment_total_count
      end

      result
    end

    def increment_total_count
      @total_count += 1
    end

    def record_failure
      @last_failure_time = Time.now
      @failure_count += 1
    end

    def invoke_callback(result = nil, options = {})
      @callback.respond_to?(:call) && @callback.call(result, options)
    end
  end
end


lib/circuit_breaker.rb
require 'circuit_breaker/version'
require 'timeout'

class CircuitBreaker
  class Open < StandardError; end

  include Version
  FAILURE_THRESHOLD = 10
  FAILURE_THRESHOLD_PERCENTAGE = 0.5
  INVOCATION_TIMEOUT = 10
  RETRY_TIMEOUT = 60

  module States
    OPEN = :open
    CLOSED = :closed
    HALF_OPEN = :half_open
  end

  attr_reader :invocation_timeout, :failure_threshold, :failure_threshold_percentage, :total_count, :failure_count

  def initialize(**options)
    @failure_count = 0
    @total_count = 0
    @failure_threshold = options[:failure_threshold] || FAILURE_THRESHOLD
    @failure_threshold_percentage = options[:failure_threshold_percentage] || FAILURE_THRESHOLD_PERCENTAGE
    @invocation_timeout = options[:invocation_timeout] || TIMEOUT
    @retry_timeout = options[:retry_timeout] || RETRY_TIMEOUT
  end

  def call(&block)
    case prev_state = state
    when States::CLOSED, States::HALF_OPEN
      connect(&block)
      update_total_count(prev_state)
    when States::OPEN
      raise Open
    end
  end

  private

  def state
    case
    when reached_failure_threshold? && reached_retry_timeout?
      States::HALF_OPEN
    when reached_failure_threshold?
      States::OPEN
    else
      States::CLOSED
    end
  end

  def reached_failure_threshold?
    (failure_count >= failure_threshold) &&
      (total_count != 0 &&
        (failure_count.to_f / total_count.to_f) >= failure_threshold_percentage)
  end

  def reached_retry_timeout?
    (Time.now - @last_failure_time) > @retry_timeout
  end

  def reset
    @failure_count = 0
    @state = States::CLOSED
  end

  def connect(&block)
    begin
      Timeout::timeout(invocation_timeout) do
        block.call
      end
      reset
    rescue Timeout::Error => e
      record_failure
    end
  end

  def update_total_count(state)
    if state == States::HALF_OPEN
      @total_count = 0
    else
      @total_count += 1
    end
  end

  def record_failure
    @last_failure_time = Time.now
    @failure_count += 1
  end
end
